subprojects/language-java/src/integTest/groovy/org/gradle/api/tasks/JavaToolchainBuildOperationsIntegrationTest.groovy

```diff
@@ -0,0 +1,429 @@
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.gradle.api.tasks
+
+import org.gradle.api.internal.tasks.execution.ExecuteTaskBuildOperationType
+import org.gradle.integtests.fixtures.AbstractIntegrationSpec
+import org.gradle.integtests.fixtures.AvailableJavaHomes
+import org.gradle.integtests.fixtures.BuildOperationsFixture
+import org.gradle.internal.jvm.Jvm
+import org.gradle.internal.jvm.inspection.JvmInstallationMetadata
+import org.gradle.internal.operations.trace.BuildOperationRecord
+import org.gradle.jvm.toolchain.internal.operations.JavaToolchainUsageProgressDetails
+import org.gradle.test.fixtures.file.TestFile
+import org.gradle.util.internal.ToBeImplemented
+
+class JavaToolchainBuildOperationsIntegrationTest extends AbstractIntegrationSpec {
+
+    def operations = new BuildOperationsFixture(executer, temporaryFolder)
+
+    def setup() {
+        buildFile << """
+            apply plugin: "java"
+
+            ${mavenCentralRepository()}
+            dependencies {
+                testImplementation 'junit:junit:4.13'
+            }
+        """.stripIndent()
+    }
+
+    @ToBeImplemented("All cases are supported except up-to-dateness for the javadoc task")
+    def "emit toolchain usages for a build #configureToolchain configured toolchain for '#task' task"() {
+        JvmInstallationMetadata jdkMetadata
+        if (configureToolchain == "without") {
+            jdkMetadata = AvailableJavaHomes.getJvmInstallationMetadata(Jvm.current())
+        } else {
+            jdkMetadata = AvailableJavaHomes.getJvmInstallationMetadata(AvailableJavaHomes.differentJdk)
+
+            if (configureToolchain == "with java plugin") {
+                configureToolchainViaJavaPlugin(jdkMetadata)
+            } else if (configureToolchain == "with per task") {
+                configureToolchainPerTask(jdkMetadata)
+            } else if (configureToolchain == "with java plugin and per task") {
+                configureToolchainViaJavaPlugin(AvailableJavaHomes.getJvmInstallationMetadata(Jvm.current()))
+                configureToolchainPerTask(jdkMetadata)
+            } else {
+                throw new IllegalArgumentException("Unknown configureToolchain: " + configureToolchain)
+            }
+        }
+
+        file("src/main/java/Foo.java") << """
+            /**
+             * This is a {@code Foo} class.
+             */
+            public class Foo {}
+        """.stripIndent()
+
+        file("src/test/java/FooTest.java") << """
+            public class FooTest {
+                @org.junit.Test
+                public void test() {}
+            }
+        """.stripIndent()
+
+        def taskPath = ":$task"
```

Comment:
Would it make sense to pass in the task path as `path` directly in the where clause, so we don't need to have the second variable in the test?